### PR TITLE
Course data exports additional features

### LINKED ISSUE #3854: Remove deactivated students from Final grade download

### PR DESCRIPTION
**Pull request by adishy**
### Status
**READY**

### Description
* Adds a column in the "Assignment Structure" export that describes the student visibility of the assignment
* In the "Final Grades" export, exports only activated students

### Migrations
NO

### Steps to Test or Reproduce
1. Visit the Course Data Exports page (i.e. /downloads) as an instructor and download the export "Final Grades". Only activated students will be exported.
2. Visit the Course Data Exports page (i.e. /downloads) as an instructor and download the export "Assignment Structure". There will be an additional column that describes the student visibility of the assignment.

### Impacted Areas in Application
* Final Grades exporter (exporters/course_grade_exporter.rb)
* Assignment Structure exporter (exporters/assignment_exporter.rb)

======================
Closes #3854, #3978

### CODE REVIEW

File: app/exporters/assignment_exporter.rb [hunk 1]
```diff
@@ -49,4 +50,18 @@ def formatted_date(date)
     return nil if date.nil?
     date.in_time_zone(@user.time_zone).strftime("%m/%d/%Y")
   end
+
+  def visibility(assignment)
+    if assignment.unlock_conditions.length > 0
```
Comment (ebarroso7) [inline app/exporters/assignment_exporter.rb]: I would prefer that this line uses `assignment.unlock_conditions.exists?` rather than the `.length > 0`

Comment (adishy) [reply app/exporters/assignment_exporter.rb]: Just pushed the change :)

File: app/exporters/course_grade_exporter.rb [hunk 2]
```diff
@@ -4,7 +4,7 @@ class CourseGradeExporter
   def final_grades_for_course(course)
     CSV.generate do |csv|
       csv.add_row baseline_headers
-      course.students.order_by_name.each do |student|
+      course.students.where(activation_state: "active").order_by_name.each do |student|
```
Comment (ebarroso7) [inline app/exporters/course_grade_exporter.rb]: making a note here to investigate activation_state some more... looking on prod there are three: "active", "pending", and nil. 
I'm also not sure where this is used and if it addresses the issue in #3854.

Comment (adishy) [reply app/exporters/course_grade_exporter.rb]: Updated to use the course membership of the student to check whether they have been deactivated in the course.

File: app/exporters/course_grade_exporter.rb [hunk 3]
```diff
@@ -4,7 +4,7 @@ class CourseGradeExporter
   def final_grades_for_course(course)
     CSV.generate do |csv|
       csv.add_row baseline_headers
-      course.students.order_by_name.each do |student|
+      course.students.order_by_name.reject{ |student| !student.course_memberships.find_by(course_id: course.id).active }.each do |student|
```
Comment (ebarroso7) [inline app/exporters/course_grade_exporter.rb]: @adishy good fix, but i still think we can do better.

I'm thinking we make this its own method, or scope -- something like `active_students`

so you can make the line as clean as `course.active_students.order_by_name.each do |student|`

Comment (adishy) [reply app/exporters/course_grade_exporter.rb]: Good idea! Will change it to use this :)

Comment (adishy) [reply app/exporters/course_grade_exporter.rb]: An active_students scope has been added in models/user.rb to get active students in a course.

File: app/exporters/assignment_exporter.rb [hunk 4]
```diff
@@ -49,4 +50,18 @@ def formatted_date(date)
...
     date.in_time_zone(@user.time_zone).strftime("%m/%d/%Y")
   end
+
+  def visibility(assignment)
+    if assignment.unlock_conditions.exists?
+      if !assignment.visible_when_locked
+        return "If Unlocked"
+      end
+      
+      return "Yes (even if assignment locked)"
```
Comment (ebarroso7) [inline app/exporters/assignment_exporter.rb]: probably more clear to say "Yes, even when locked"

File: app/exporters/assignment_exporter.rb [hunk 5]
```diff
@@ -49,4 +50,18 @@ def formatted_date(date)
     return nil if date.nil?
     date.in_time_zone(@user.time_zone).strftime("%m/%d/%Y")
   end
+
```
Comment (ebarroso7) [inline app/exporters/assignment_exporter.rb]: I'm thinking the column options should be like:
"Visible when unlocked"
"Visible when locked"
"Visible"
"Hidden"

Comment (ebarroso7) [reply app/exporters/assignment_exporter.rb]: changed my mind!! Good night sleep always does it...
"Hidden when locked"

Comment (adishy) [reply app/exporters/assignment_exporter.rb]: This has been updated as specified :)

**Pull request merged**